lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.reef.driver.context.ActiveContext;
+import org.apache.reef.driver.context.ClosedContext;
+import org.apache.reef.driver.context.ContextBase;
+import org.apache.reef.tang.ClassHierarchy;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Factory for ActiveContextBridge instances.
+ * <p/>
+ * This class is thread safe because all its methods are synchronized. It should only be used within the bridge code.
+ */
+@DriverSide
+@ThreadSafe
+@Private
+public final class ActiveContextBridgeFactory {
+  private static final Logger LOG = Logger.getLogger(ActiveContextBridge.class.getName());
+
+  private final Map<String, ActiveContextBridge> activeContextBridgeCache = new HashMap<>();
+  private final ClassHierarchy clrClassHierarchy = Utilities.loadClassHierarchy(NativeInterop.CLASS_HIERARCHY_FILENAME);
+  private final AvroConfigurationSerializer configurationSerializer;
+
+  /**
+   * This is always instantiated via Tang.
+   *
+   * @param configurationSerializer passed to the ActiveContextBridge instances for configuration serialization.
+   */
+  @Inject
+  private ActiveContextBridgeFactory(final AvroConfigurationSerializer configurationSerializer) {
+    this.configurationSerializer = configurationSerializer;
+  }
+
+  /**
+   * Instantiates a new ActiveContextBridge or returns the one created earlier.
+   *
+   * @param context the context for which to return an ActiveContextBridge instance.
+   * @return a new ActiveContextBridge or returns the one created earlier.
+   */
+  public synchronized ActiveContextBridge getActiveContextBridge(final ActiveContext context) {
+    final String contextId = getUniqueContextId(context);
+
+    ActiveContextBridge result = activeContextBridgeCache.get(contextId);
+    if (null == result) {
+      result = new ActiveContextBridge(context, this.clrClassHierarchy, this.configurationSerializer);
+      activeContextBridgeCache.put(contextId, result);
+    }
+    return result;
+  }
+
+  /**
+   * Removes the ActiveContextBridge associated with the given Context from the cache.
+   *
+   * @param context the context whose ActiveContextBridge instance shall be removed from the cache.
+   */
+  public synchronized void ForgetContext(final ClosedContext context) {
+    final String contextId = getUniqueContextId(context);
+    final ActiveContextBridge removed = this.activeContextBridgeCache.remove(contextId);
+    if (null != removed) {
+      LOG.warning("Called with context previously unknown: " + contextId);
+    }
+  }
+
+  /**
+   * Constructs a unique ID for the given context by combining the evaluator and context ids.
+   *
+   * @param contextBase
+   * @return
+   */
+  private static String getUniqueContextId(final ContextBase contextBase) {
+    return contextBase.getEvaluatorId() + "/" + contextBase.getId();
```

Comment:
I think we may need to take a step back to examine if making fundamental changes such as Option 3 to reef framework is worthy for the purpose of this PR. 

I recalled we had this discussion a while back as why it is a good practice to have user assign the contextId, instead of generating one (unique or not) from within reef framework: the reason is that user want to know that the context he just requested has been allocated and returned (in acticecontexthandler), and that is done by the user-defined contextId (same for taskId). This is a very helpful feature when user want to run different things in different contexts with different (user) settings. Sure we lost some control over the choices of contextId, but that is well worth it, IMHO. I think we had this problem with evaluatorId at some point: since user has no control over eveluatorId, he cannot be sure whether the evaluator he got is the right one (with some specific resource) and we need to pull a pretty niffy trick to resolve that. So, I would prefer that we leave the contextId/taskId to users. Unfortunately that would actually rule out both Option 2 and Option 3.

If we really want to guarantee id uniqueness, we can do option 1, that is, let user provide contextId, but reef will complain when the contextId is not unique (at runtime). 

Or rather, just don't cache the context (we can still guarantee single copy of classhierarchy and searilzer through the factory), new up a new context whenever needed, as I mentioned in the note I added earlier. After all that is not a lot of overhead (both memory and performance wise).

my 2 cents.